### PR TITLE
Fix to use semantic function default values

### PR DESCRIPTION
**Pull request by markwallace-microsoft**
### Motivation and Context

Please help reviewers and future users, providing the following information:
  1. Why is this change required? Default values specified in a semantic function config.json are not being used when executing the semantic function

### Description

1. Add logic to use default values when executing a semantic function
2. Add a new extension method to allow execution of the text prompt

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/SemanticKernel/SkillDefinition/SKFunction.cs [hunk 1]
```diff
@@ -264,8 +264,12 @@ public async Task<SKContext> InvokeAsync(SKContext context, CompleteRequestSetti
 
         if (this.IsSemantic)
         {
+            var addedVariables = this.AddDefaultValues(context.Variables);
+
             var resultContext = await this._function(this._aiService?.Value, settings ?? this._aiRequestSettings, context).ConfigureAwait(false);
             context.Variables.Update(resultContext.Variables);
+
+            this.RemoveDefaultValues(context.Variables, addedVariables);
```
Comment (dmytrostruk) [inline dotnet/src/SemanticKernel/SkillDefinition/SKFunction.cs:272]: I think it's a little bit dangerous to remove something from the context after function execution based on function configuration.

Scenario:
_SKFunction1_ has `input` parameter with default value, value is not presented by user.
_SKFunction2_ has `input` parameter without default value, which should be taken from _SKFunction1_.

Action:
I call two functions in pipeline - `kernel.RunAsync(skFunction1, skFunction2)`

Expected behavior:
_SKFunction1_ returns result and put it in SKContext as `input` for next _SKFunction2_.
_SKFunction2_ function will use `input` parameter with response from _SKFunction1_.

Actual behavior:
_SKFunction1_ adds default value to the context and executes. In `addedVariables` we will have one parameter `input`. After execution it removes `input` from SKContext.
_SKFunction2_ has no default parameter value (as described in Scenario) and input from previous _SKFunction1_ was removed. So _SKFunction2_ will be without `input`.

I may have missed something, let me know what you think about this case. Maybe worth to check this behavior with unit-tests.

File: dotnet/src/SemanticKernel/SkillDefinition/SKFunction.cs [hunk 2]
```diff
@@ -1033,5 +1040,29 @@ private static void TrackUniqueParameterType(ref bool hasParameterType, MethodIn
     /// <summary>Formatter functions for converting parameter types to strings.</summary>
     private static readonly ConcurrentDictionary<Type, Func<object?, CultureInfo, string>?> s_formatters = new();
 
+    /// <summary>Add default values to the context variables if the variable is not defined</summary>
+    private List<string> AddDefaultValues(ContextVariables variables)
+    {
+        var addedVariables = new List<string>();
+        foreach (var parameter in this.Parameters)
+        {
+            if (!variables.ContainsKey(parameter.Name) && parameter.DefaultValue != null)
```
Comment (dmytrostruk) [inline dotnet/src/SemanticKernel/SkillDefinition/SKFunction.cs:1049]: Should we also handle the case when parameter exists in context variables, but it's null or empty?

File: dotnet/src/SemanticKernel/SkillDefinition/SKFunction.cs (hunk 1 above)
Comment (markwallace-microsoft) [reply dotnet/src/SemanticKernel/SkillDefinition/SKFunction.cs]: Good suggestion.

File: dotnet/src/SemanticKernel/SkillDefinition/SKFunction.cs (hunk 2 above)
Comment (markwallace-microsoft) [reply dotnet/src/SemanticKernel/SkillDefinition/SKFunction.cs:1043]: The current thinking is only use defaults if the parameter has not been set. Depending on the prompt null or empty string could be expected values.

@shawncal what do you think?

Comment (shawncal) [reply dotnet/src/SemanticKernel/SkillDefinition/SKFunction.cs:1043]: Yeah, I'd prefer to use default value if `null` (not set). If user has set value to `string.Empty`, that's a valid, explicit choice that could be a way of saying "ignore this value -- add nothing to the prompt here" and a way of inserting optional clauses into a prompt.

**Review by shawncal — APPROVED** [review]

**Pull request merged**